A different approach to scatter/gather

tfenne opened the pull request.

tfenne commented:
@nh13 I forked your branch and took a shot at implementing the other scatter/gather API we talked about.  I translated your pipeline to use the new API and after a little testing and bug-fixing it works!  The code behind the API is fairly short, but very confusing and I have a hard time keeping it straight!

Next steps are:
1. Review with you to make sure you still prefer this API
2. Convert all the tests and maybe write some new ones too
3. Fix the fact that, for now, `a ==> Scatter` and `Scatter ==> b`, where `a` and `b` are external/unrelated tasks, doesn't work

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+   * A type of Task that, when run, partitions some kind of input, and produces one or more outputs
+   * of type A that should be parallelized (scattered) over.
+   */
+  trait Scatterer[A] extends Task {
```
Just like a `Pipe`, I would prefer to have the input type and output type be present.   For example, `Scatterer[PathToVcf, Seq[VariantContext]]` or `Scatterer[PathToInveralList, Seq[IntervalList]]` or `Scatterer[SequenceDictionary, ReferenceSequence]`.  This tells the reader what the input is that is being scattered and possibly transformed.  I mention the latter since it isn't always an input of type `A` to an output of type `Seq[A]`.

  On the other hand, one could argue that it really doesn't matter what the inputs are to a `Scatterer`, just that it produces a sequence of outputs of a fixed type.  Thoughts?

tfenne replied:
I sort of agree, that it would be kinda be nice documentation to require `Scatterer`s to declare their input and output types, I decided against it initially because it's not necessary and since this is the one trait that tasks do have to implement, I wanted to keep it as light as possible.  `Scatterer` is really a `Producer[A]`, it's role is to produce partitions over a domain, but we don't really have an opinion about how it does so.  

For example if there are multiple inputs - say a set of intervals and a BAM and a reference - and all are used to compute the output set of intervals, would that be a `Scatterer[(PathToBam, PathToFasta, PathToIntervals), PathToIntervals]`?  Or just a `Scatterer[PathToIntervals,PathToIntervals]`?

Perhaps it's the name that's the problem?  Scatterer does imply taking something and scattering it - perhaps it should be `PartitionProducer` or something else?

nh13 replied:
I just ask myself what does it scatter?  Perhaps look to other software that implements scatter/gather or map/reduce to find an appropriate name.

tfenne replied:
A little googling didn't turn up much.  Queue isn't helpful (`ScatterFunction` and `GatherFunction`), and there's just not much else out there.  I think probably it does make sense to call it `PartitionProducer` or `PartitionMaker` or something similar.

nh13 replied:
I like the one word name and something short: `Splitter`, `Divider`, `Separator`, `Subdivider`.  We can keep `Scatterer` for now.

As per your input example (`Scatterer[(PathToBam, PathToFasta, PathToIntervals)`), I would say pick what is the primary input data source, and this case likely the `PathToIntervals`.  While not ideal, I don't think we want ``Scatterer[(PathToBam, PathToFasta, PathToIntervals, Int, Int, Int, Int, Float, Float)` where we include all of the parameters to the scatterer.  I think this is more subjective and meant for clarity, rather than a rule.

tfenne replied:
I'm actually tempted to rename it `Partitioner` then, since even though I wrote it I find it a little hard to visually distinguish between `Scatterer` and `Scatter`

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+
```
I'll forgive you this time for such a blatant error.

tfenne replied:
Hah - this is actually on _purpose_!  When I have lots of little classes and traits together in a file I like to have a couple of blank lines between them to make it easier to see where one ends and the next one begins.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+  /** Provides a factory method to create a Scatter object from a Scatterer. */
+  object Scatter {
+    def apply[A](scatterer: Scatterer[A]): Scatter[A] = {
+      new ScattererWrapper(scatterer)
```
one liner?

tfenne replied:
yup

nh13 commented on pipelines/src/main/scala/dagr/pipelines/ParallelHaplotypeCaller.scala:
```diff
@@ -101,14 +60,29 @@ class ParallelHaplotypeCaller
+    }
+
+    root ==> scatter
+    gather ==> new DeleteFiles(dir)
```
It would look better if it was `root ==> scatter ==> gather ==> new DeleteFiles(dir)`.   Also, the files aren't delete as soon as they could be, but this likely would be fine.

tfenne replied:
I think this would be fine - though actually it would be:

``` scala
root ==> scatter ==> hc ==> gt ==> gather ==> new DeleteFiles(dir)
```

It's just that `scatter ==> hc ==> gt ==> gather` is auto-wired by the scatter/gather mechanics.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+   * Represents a gather operation in a scatter/gather pipeline.
+   */
+  sealed trait Gather[A, B <: Task] extends Task {
+    private[ScatterGather] var sources: Option[Seq[A]] = None
```
Could we scatter a `Gather`?  For example, we gather to a `Path` and then want to scatter again. 

```
/**
 * Produces a new Scatter from the gathered task.  Takes a function that maps an Task of type B to a Scatter of object of type C. 
 * 
 * The resulting Scatter[C] can be further mapped or gathered (or both!).
 */
def scatter[C](f: B => scatterer[C]): Scatter[C] = throw new NotImplementedException()
```

This is close to using a `Linker` task.

tfenne replied:
We could do this, but I'm not sure how that's really different than just doing:

``` scala
val gather = scatter.gather(things => new GatherThings(in=things, out=gatheredThing)
val nextScatter = Scatter(new ThingScatterer(in=gatheredThing))
```

nh13 replied:
`gatheredThing` needs to be known ahead of time in your example, I don't like that.

tfenne replied:
That's only partly true.  More specifically the place where `GatherThings` is going to deposit it's result needs to be known ahead of time.  `gatheredThing` could easily be a simple container object into which `GatherThings` places something (e.g. an `AtomicReference[Thing]`).  But I see your point.  Let's log an issue for this and implement it when we need it?  You're right that's it's totally possible.

nh13 replied:
Lets defer when we have an example, but food for thought.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+    }
+  }
+
+
```
👎 nl

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+  * Implementation of a Scatter that is just a thinly veiled wrapper around the Scatterer
+  * being used to generate the set of scatters/partitions to operate on.
+  */
+  private class ScattererWrapper[A](scatterer: Scatterer[A]) extends Scatter[A] {
```
why not put this in the `Scatter` object.  I found it hard to find and it is only used there.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+    /** Ensures that gathers are wired in correctly. */
+    protected def connectGathers: Unit = {
+      gathers.foreach(gather => {
+        gather.sources = Some(tasks)
```
`tasks.toSeq`?

tfenne replied:
Intellij tells me that is a no-op.  I guess tasks.toList would be most defensive - I can do that.

nh13 replied:
Yeah, just in case it's a silly iterator.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+   * of type A that should be parallelized (scattered) over.
+   */
+  trait Scatterer[A] extends Task {
+    def scatters: Seq[A]
```
Should `scatters` return `Option[Seq[A]]`?  This would allow the `ScatterGather` to detect when `scatters` is called too early, and can then thrown an exception, rather than having each implementation do the guarding.

tfenne replied:
Sure - I went back and forth on this.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+  * scatterer has created the partitions.  It's tasks are generated using the scatter partitions
+  * as input, and is also responsible for chaining together sub-scatters.
+  */
+  private class PrimarySubScatter[A, B <: Task](val scatterer: Scatterer[A], f: A => B) extends Pipeline with SubScatter[A,B] {
```
I don't think `scatterer` needs to be a `val`, since it is only access inside this class.

tfenne replied:
I've made it a `private val`.  FWIW if you access a constructor parameter anywhere other than in the constructor, scala promotes it to a val anyway, so I prefer to declare it that way.

nh13 commented on tasks/src/main/scala/dagr/tasks/ScatterGather.scala:
```diff
@@ -0,0 +1,175 @@
+  */
+  private trait SubScatter[A, B <: Task] extends Scatter[B] {
+    val tasks   = new ListBuffer[B]
+    val subs    = new ListBuffer[SubScatter[B, _ <: Task]]
```
Could this just be the following?
`val subs    = new ListBuffer[SecondarySubScatter[B, _ <: Task]]`
It may make it clearer that the `PrimarySubScatter` is the one that takes the initial scattered inputs while the `SecondarySubScatter` tasks simply chain along the "scatter" prior to a final gather.

tfenne replied:
Yes!  Thanks for catching that.  At one point with the way I had things this didn't work, but it does now and it does clean things up a bit.

nh13 commented:
@tfenne I did a thorough review (needs tests!) since once I understood what you were doing, I think you were close enough to warrant it.  Perhaps an explanation at the top of the source file showing an example of how all these classes work together at a high-level would be nice too to remind our future selves.  

My only discussion point was around whether or not we should have `Scatterer[A]` or `Scatterer[A,B]`.

tfenne commented:
@nh13 I've added a  decent test, cleaned up a touch and squashed.  I don't want to add the extra type parameter to `Partitioner` since it's not needed I think it feels like ceremony.  Want to take a final look before I merge?

nh13 commented on pipelines/src/main/scala/dagr/pipelines/ParallelHaplotypeCaller.scala:
```diff
@@ -0,0 +1,88 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016 Fulcrum Genomics
```
LLC?

nh13 commented on tasks/src/test/scala/dagr/tasks/ScatterGatherTests.scala:
```diff
@@ -0,0 +1,114 @@
+  * Our test involves a "complicated" little scatter/gather that does the following:
+  *   - Scatter a text file by lines
+  *   - Count the words on each line
+  *   - [Gather] Square the counts
```
perhaps "square each count" to "differentiate it from "sum the counts", where the former is a map and the latter is a gather?

nh13 commented on tasks/src/test/scala/dagr/tasks/ScatterGatherTests.scala:
```diff
@@ -0,0 +1,114 @@
+  *   - Count the words on each line
+  *   - [Gather] Square the counts
+  *   - Sum the counts
+  *   - [Gather] Sum the squares
```
this description doesn't jive with what I see below.  How about

```
- [Scatter] Scatter the text file by lines
- [Map] Count the words on each lines
- [Map] Square the count for each line
- [Gather] Sum the counts across all lines.
```

?

nh13 replied:
Probably need a test or two that does:
- `Scatter, Map, Gather, Scatter, Map, Gather`
- `Scatter, Map, Gather.  Map on the original Scatter, and another Gather` (is this supported)?

You could log an issue and I'd be happy to take a shot at adding those in (after you merge this PR).

tfenne replied:
On your first comment here - I accidentally a couple of words.  Square and Sum got swapped.  Fixed now.

On your second line comment here - this test actually _is_ that second case.  It scatters, maps to counts and gathers the counts, and then in parallel maps the `counts` to `counts^2` and gathers those.  Or is this not what you meant?

nh13 commented:
minor stuff, read for merge at your discretion, @tfenne back to you.

tfenne commented:
Fixed the minor points.  Let's get this merged and then add more test cases as we need them.

nh13 commented:
+1

nh13 commented:
Thanks!
